src/include/common/dedicated_thread_owner.h

```diff
@@ -16,35 +18,65 @@ namespace terrier {
  */
 class DedicatedThreadOwner {
  public:
+  virtual ~DedicatedThreadOwner() = default;
   /**
    * @return the number of threads owned by this owner
    */
-  size_t GetThreadCount() { return thread_count_; }
+  size_t GetThreadCount() {
+    std::unique_lock<std::mutex> lock(thread_count_latch_);
+    return thread_count_;
+  }
+
+  /**
+   * Notifies the owner that a new thread can be given to it. The thread owner has the opportunity to decline the new
+   * thread if it does not need it
+   * @warning Should only be called by self driving infrastructure.
+   * @return true if owner accepts thread, false if it declines it
+   */
+  bool NotifyNewThread() { return OnThreadGranted(); }
 
   /**
    * Notifies the owner that a new thread has been given to it
    */
-  void NotifyNewThread() { thread_count_++; }
+  void GrantNewThread() {
+    std::unique_lock<std::mutex> lock(thread_count_latch_);
+    thread_count_++;
+  }
 
   /**
-   * Notifies the owner that the thread running task will be terminated
+   * Notifies the owner that the thread running task will be terminated. The owner has the opportunity to reject the
+   * removal of the thread if it needs it
    * @param task the task to be terminated
+   * @return true if owner accepts thread removal, false if it rejects
    */
-  void NotifyThreadRemoved(const std::shared_ptr<DedicatedThreadTask> &task) {
+  bool NotifyThreadRemoved(common::ManagedPointer<DedicatedThreadTask> task) { return OnThreadRemoved(task); }
+
+  /**
+   * Notifies the owner that a new thread has removed from them
+   */
+  void RemoveThread() {
+    std::unique_lock<std::mutex> lock(thread_count_latch_);
     thread_count_--;
-    OnThreadRemoved(task);
   }
 
  protected:
+  /**
+   * Custom code to be run when offered a thread by each owner. The owner has the option to decline the new thread if it
+   * does not need it. If the owner accepts, its up to the owner to call RegisterDedicatedThread to register their task
+   * @return true if owner accepts thread, false if it declines it
+   */
+  virtual bool OnThreadGranted() { return false; }
```

Comment:
Unless we merge the two public methods above, having this be separate from `NotifyNewThread` doesn't buy us much.